### ISSUE TITLE
Cookie does not get set on Firebase App Hosting

### ISSUE THREAD

**Issue by gabor-s** [awinogrodzki/next-firebase-auth-edge#188]
We are evaluating [Firebase App Hosting](https://firebase.google.com/docs/app-hosting), but as the title says the cookie does not get set. We can see the message `next-firebase-auth-edge: Updating response headers with authenticated cookies` in the logs, but there's no `Set-Cookie` header in the response. The cookie is named `__session`. Locally and on Vercel it's working fine, the cookie get set.

**Comment by abrahan92** [comment]
Yes, I have this issue as well. Even using the cookie: `__session` is not being set/cached on the response.

I've tried using the last pre-release canary version but didn't have any luck

`"next-firebase-auth-edge": "^1.5.0-canary.4",`

I'm running nextJS app on firebase App Hosting

```
return authMiddleware(request, {
    loginPath: '/api/login',
    logoutPath: '/api/logout',
    apiKey: authConfig.apiKey,
    cookieName: authConfig.cookieName,
    cookieSerializeOptions: authConfig.cookieSerializeOptions,
    cookieSignatureKeys: authConfig.cookieSignatureKeys,
    serviceAccount: authConfig.serviceAccount,
    handleValidToken: async ({ token, decodedToken }, headers) => {
      console.log('Token is valid')

      if (publicRoutes.includes(request.nextUrl.pathname)) {
        return redirectToHome(request)
      } else if (request.nextUrl.pathname === '/') {
        url.pathname = '/home'

        return NextResponse.redirect(url)
      }

      return NextResponse.next({
        request: {
          headers
        }
      })
    },
    handleInvalidToken: async reason => {
      console.info('Missing or malformed credentials', { reason })

      return redirectToLogin(request, {
        path: '/login',
        publicPaths: publicRoutes
      })
    },
    handleError: async error => {
      console.error('Unhandled authentication error', { error })

      return redirectToLogin(request, {
        path: '/login',
        publicPaths: publicRoutes
      })
    }
  ```

and this is my cookieSerializeOptions 

```
cookieSerializeOptions: {
    path: '/',
    httpOnly: true,
    secure: process.env.NEXT_PUBLIC_USE_SECURE_COOKIES === 'true',
    sameSite: 'lax',
    maxAge: 12 * 60 * 60 * 24
  }
```

**Comment by abrahan92** [comment]
I'm deploying directly from the Firebase App Hosting, selecting a specific branch, not the Firebase CLI. But even the cloud-run server is running well, without issues. Also, I'm using the env variables from Google Secret Manager.

I have a middleware.js within my nextJS /src folder, which works locally. The thing is, on production does not set the cookie even when it is called `__session`

These are the production logs within the cloud-run server
![image](https://github.com/awinogrodzki/next-firebase-auth-edge/assets/13084167/45b9cc08-7dc7-42cd-bd0a-4acdb7f6a7fd)


This is where I make the deployments. **Firebase AppHosting deploys to Cloud Run**

![image](https://github.com/awinogrodzki/next-firebase-auth-edge/assets/13084167/9705cd0d-95d9-405d-88b6-158a283cefca)

**Comment by awinogrodzki** [comment]
@abrahan92 thanks for the logs.

In the logs, I cannot see calls to GET `/api/login` endpoint. Could you open browser dev tools, go to the Network tab and copy all the details of `GET /api/login` request/response and share it here?

**Comment by abrahan92** [comment]
This is the response from the `/api/login` endpoint by the `authMiddleware`

![image](https://github.com/awinogrodzki/next-firebase-auth-edge/assets/13084167/f8277c62-b62b-47f0-898c-9588cc585850)

Also, here you can see that I'm using the correct cookie `__session` as I get from env vars

![image](https://github.com/awinogrodzki/next-firebase-auth-edge/assets/13084167/5d20027d-5fd9-4443-a76b-2cfb36d3180e)

**Comment by gabor-s** [comment]
Hi @awinogrodzki 
We used version `1.4.5`, now I tried `1.5.0-canary.4`, but it's the same.

Options:
![image](https://github.com/awinogrodzki/next-firebase-auth-edge/assets/870387/06220adf-b1d9-4a0f-87fa-ce286cb726a2)

Request/response:
![image](https://github.com/awinogrodzki/next-firebase-auth-edge/assets/870387/42358551-eac8-4f31-b375-5f5b90167929)

**Comment by gabor-s** [comment]
Firebase App Hosting and Firebase Hosting are two different beasts. For Firebase Hosting it's documented that the cookie name should be `__session`, but I couldn't find any information regarding Firebase App Hosting and cookies.

**Comment by abrahan92** [comment]
Me neither; I assumed that worked similarly with Cloud Run, but it seems it doesn't

**Comment by awinogrodzki** [comment]
I have reproduced the issue in App Hosting environment, here: https://next-firebase-auth-edge--next-firebase-auth-edge-demo.us-central1.hosted.app/login?redirect=/

It seems App Hosting strips Set-Cookie headers from /api/login calls. I'll make some more experiments and get back to you

**Comment by awinogrodzki** [comment]
@abrahan92 @gabor-s 

This seems like an issue with App Hosting itself.

Here's App Hosting version of the app: https://next-firebase-auth-edge--next-firebase-auth-edge-demo.us-central1.hosted.app/login?redirect=/


And here's the same instance, accessed by Cloud Run url directly: https://next-firebase-auth-edge-m2gwdqnpta-uc.a.run.app/

When you open Dev Tools > Network tab to analyze requests, you can see that in Cloud Run environment we have following headers attached to each request:
```
Set-Cookie:
X-Test-Cookie=test-value; Max-Age=1036800; Path=/; HttpOnly; Secure; SameSite=Lax
Set-Cookie:
X-Test-Cookie-2=test-value-2
```

When you use App Hosting url, `Set-Cookie` headers are stripped by `envoy` client. 

I cannot find a way to overwrite this behaviour in App Hosting settings. This is a good candidate to report to Google team

**Comment by abrahan92** [comment]
Hey @awinogrodzki 

Having spent more than five days testing and even creating a combination of your library and cloud functions, I won't be able to fix that behavior. That's what I feared.

It's worth reporting it to them. I will

The first one to be served can come and let us know.

Thanks both

**Comment by minhdanh** [comment]
Hi @gabor-s, did you have this fixed?

**Comment by gabor-s** [comment]
Hi @minhdanh , nope, their "engineering team is still discussing" the issue.

**Comment by gabor-s** [comment]
I've just found this in Firebase App Hosting [Known issues and limitations](https://firebase.google.com/docs/app-hosting/about-app-hosting#known-issues):
> Set-Cookie headers are stripped from responses served through the App Hosting dataplane. A fix will be available in a later release.

**Comment by discoverlance-com** [comment]
Yes, it looks like unless they resolve it, it makes the package unusable there.

I found a workaround where you can go to the cloud run service that is deployed for app hosting and then Under Integrations, you can go to Add Integration and then add Firebase Hosting. You can then choose a suitable site ID which will allow you to integrate firebase hosting with the cloud run service.
![image](https://github.com/awinogrodzki/next-firebase-auth-edge/assets/52635726/7995d730-9138-44eb-825f-1de4f78e3159)

NOTE:
1. You will have to allow unauthenticated access to your cloud run service. App Hosting enables require authentication on your cloud run service which does not work with Firebase Hosting so you can go to the security tab on cloud run and update that.
![image](https://github.com/awinogrodzki/next-firebase-auth-edge/assets/52635726/eebc7bc2-d948-4c98-9510-9c4da94eb6e8)

2. Change your session cookie name to `__session` as firebase hosting also has the requirement of having `__session` as the cookie name for it to work.
```ts
export const authConfig = {
	apiKey: serverConfig.firebaseApiKey,
	cookieName: '__session',
        ...
}

```

3. If you have set an environment variable like `NEXT_PUBLIC_APP_URL` in your **apphosting.yaml** file, you will have to update that to the new site id url you obtained after enabling fireabase hosting.

You can then access your application with the firebase hosting url and use the CI/CD from app hosting to deploy your application as usual so you still have app hosting working but you will just not use the app hosting url when you want to access your application but the firebase hosting url. At least till their team fixes app hosting cookie issue.

Since firebase hosting is also free, you can enable it and not worry about cost.  Only your cloud run/secrets or any other cost will count.

**Comment by Jacquesjh** [comment]
Could this be related to #195 ?

**Comment by awinogrodzki** [comment]
@Jacquesjh, I don't think so – this issue is about Firebase App Hosting stripping `Set-Cookie` headers from API response

Feel free to create a new issue for the problem you're experiencing

**Comment by discoverlance-com** [comment]
Firebase has updated the documentation on app hosting to remove the [known issues and limitations](https://firebase.google.com/docs/app-hosting/about-app-hosting#known-issues) but it still does not work... I am not sure if they have not fixed it, they have moved the documentation or it's something else...

**Comment by gabor-s** [comment]
In September they wrote this on their [blog](https://firebase.blog/posts/2024/09/app-hosting-environments):

> We’ve heard from many of you that getting Authentication working on App Hosting is difficult, given the current lack of support for Cookies. We’re working to address these limitations and have a more seamless experience with our SDKs in the future.

**Comment by gabor-s** [comment]
Hi, I got a response from Firebase support that the issue is resolved and indeed the Set-Cookie header is in the response when we're using the cookie name `__session` (we haven't tried with other cookie names).

We can use next-firebase-auth-edge on Firebase App Hosting.

**Comment by awinogrodzki** [comment]
Thank you for keeping us up-to-date @gabor-s! I guess we can close the issue in the coming

**Comment by cjangrist** [comment]
@gabor-s is this confirmed working?

**Comment by gabor-s** [comment]
> [@gabor-s](https://github.com/gabor-s) is this confirmed working?

Yes, it's working.
